src/collapse/xdebug.rs

```diff
@@ -0,0 +1,367 @@
+                Interned::New(new) => Interned::New(Call::WithPath(st, new)),
+            },
+            Interned::Old(other) => Interned::Old(Call::WithoutPath(other)),
+            Interned::New(new) => Interned::New(Call::WithoutPath(new)),
+        };
+
+        let idx = self.intern(new_or_not);
+        self.stack.push(idx)
+    }
+
+    /// Pop one call from the current stack.
+    pub fn pop(&mut self) {
+        self.stack.pop();
+    }
+
+    /// An empty stack means we finished main.
+    fn is_empty(&self) -> bool {
+        self.stack.is_empty()
+    }
+
+    /// Get a comparable, hashable representation of the call stack. of the call stack.
+    fn current(&self) -> &[usize] {
+        self.stack.as_slice()
+    }
+
+    /// Intern a string, return the unique index.
+    ///
+    /// `Ok` when the string was already present.
+    fn intern_str(&mut self, string: &str) -> Interned<Str> {
+        let mut hasher = self.strings.hasher().build_hasher();
+        string.hash(&mut hasher);
+        let hash = hasher.finish();
+
+        let entry = self
+            .strings
+            .raw_entry_mut()
+            .from_key_hashed_nocheck(hash, string);
+
+        let vacant = match entry {
+            RawEntryMut::Occupied(occ) => return Interned::Old(Str(*occ.get())),
+            RawEntryMut::Vacant(vacant) => vacant,
+        };
+
+        let index = self.interned_string.len();
+        let element: Rc<str> = Rc::from(string);
+        self.interned_string.push(element.clone());
+        vacant.insert_hashed_nocheck(hash, element, index);
+        Interned::New(Str(index))
+    }
+
+    /// Intern a `Call` into a `usize` index.
+    ///
+    /// Some tests revealed that this was faster than comparing the `Call` instances directly, so
+    /// it was kept. This seems a bit odd but was backed by data at the time.
+    fn intern(&mut self, call: Interned<Call>) -> usize {
+        let new = match call {
+            // The strings were not seen before, definitely new.
+            Interned::New(t) => t,
+            // The strings used were seen before, but maybe not in this call. So retest.
+            Interned::Old(t) => {
+                if let Some(idx) = self.calls.get(&t) {
+                    return *idx;
+                } else {
+                    t
+                }
+            }
+        };
+
+        let index = self.interned.len();
+        self.interned.push(new);
+        self.calls.insert(new, index);
+        index
+    }
+
+    /// Prepare the stack frame for printing.
+    fn frames<'a>(&'a self, stack: &'a [usize]) -> Frames<'a> {
+        Frames { calls: self, stack }
+    }
+
+    /// Create a name for the current stack.
+    ///
+    /// This is potentially costly.
+    fn write_name(&self, indices: &[usize], buffer: &mut fmt::Formatter) -> fmt::Result {
+        let mut indices = indices.iter().cloned();
+        if let Some(first) = indices.by_ref().next() {
```

Comment:
I think I'd just use `.peekable()` on the iterator for this, but :man_shrugging: